### PR TITLE
feat: expose types

### PR DESCRIPTION
**Pull request by dpschen**
I realised that this library uses types but they were not exposed.
By trying to fix this I went a bit overboard and did the following:

- expose types for library users
- upgrade packages and vuepress
- use vite for bundling
- provide context with new composable 'useContext'
- use prettier to format
- fix minor type issues
- styles:
   - remove style scoping, to make overwriting styles easier
   - remove usage of ids (in case you want to use the chart multiple times)
   - prefix animations
   - inherit font by default, since Helvetica isn't available on every system
   - use ew-resize cursor to show that resizing is possible in two directions and align with common ui pattern
- use correct typographic character for range values
- use numeric font variant so that dates always require the same width
- align wording 'getRowsInChart' and 'getChartRows'

Somewhere along the way I broke something minor (the rendering is a bit broken right now). If there is general interest in merging these or some of these changes I would try to look into what is creating the issue.

## TODO:
- [x] use postcss-preset-env to remove the need to manually write vendor prefixes
- [x] fix newly introduced rendering bug

### CODE REVIEW

**Comment by InfectoOne** [comment]
Hello! Yes I'd be interested in merging this. Could you just clarify the following point:
> use numeric font variant so that dates always require the same width

**Comment by dpschen** [comment]
> Hello! Yes I'd be interested in merging this. Could you just clarify the following point:
> 
> > use numeric font variant so that dates always require the same width

Sure! This is regarding https://github.com/dpschen/vue-ganttastic/blob/feature/expose-types/src/components/GGanttChart.vue#L249. By defining `font-variant-numeric: tabular-nums;` we use the OpenType feature `tnum` if the used font supports it. This makes sure that numbers always have the same width (aka optimised for tables). I added this because the time preview in the tooltip 'jumps' around when dragging the start or end of a bar.

These screens are not perfectly cut but maybe you can see the different width here in the minute numbers of the start time (`11` vs `59`). It's better visible when it is in motion while you drag it around.

SCR-20220902-sw5
SCR-20220902-svi

**Comment by InfectoOne** [comment]
> > Hello! Yes I'd be interested in merging this. Could you just clarify the following point:
> > > use numeric font variant so that dates always require the same width
> 
> Sure! This is regarding https://github.com/dpschen/vue-ganttastic/blob/feature/expose-types/src/components/GGanttChart.vue#L249. By defining `font-variant-numeric: tabular-nums;` we use the OpenType feature `tnum` if the used font supports it. This makes sure that numbers always have the same width (aka optimised for tables). I added this because the time preview in the tooltip 'jumps' around when dragging the start or end of a bar.
> 
> These screens are not perfectly cut but maybe you can see the different width here in the minute numbers of the start time (`11` vs `59`). It's better visible when it is in motion while you drag it around.
> 
> SCR-20220902-sw5 SCR-20220902-svi

Excellent! Thumbs up from me - tell me when you fix the broken rendering and I will review this pull request :)

**Comment by dpschen** [comment]
@InfectoOne: If you have time you can give it a try :)

File: src/composables/useBarDrag.ts [hunk 1]
```diff
@@ -1,108 +1,123 @@
...
+import provideConfig from "../provider/provideConfig"
+import { type Ref, ref } from "vue"
 
+import type { GanttBarObject } from "../types"
+import useDayjsHelper from "./useDayjsHelper"
 import useTimePositionMapping from "./useTimePositionMapping"
-import { Ref, ref } from "vue"
 
-export default function useBarDrag (
+export default function createBarDrag(
```
Comment (InfectoOne) [inline src/composables/useBarDrag.ts:9]: It'd be better to keep the name `useBarDrag`, since it's in accordance with Vue's [naming convention for composables](https://vuejs.org/guide/reusability/composables.html#conventions-and-best-practices) and because the file itself has the same name.

File: src/composables/useBarDrag.ts [hunk 2]
```diff
@@ -1,108 +1,123 @@
...
+      default:
+        dragCallBack = drag
     }
+    isDragging.value = true
+    window.addEventListener("mousemove", dragCallBack)
+    window.addEventListener("mouseup", endDrag)
   }
 
-  const drag = (e: MouseEvent) => {
+  function getBarElements() {
```
Comment (InfectoOne) [inline src/composables/useBarDrag.ts:49]: Good idea to wrap this in a helper function! How about converting in to an arrow function so that it's consistent with other functions?

File: src/composables/useBarDrag.ts [hunk 3]
```diff
@@ -1,108 +1,123 @@
...
+    if (toDayjs(newBarEnd).isSameOrBefore(toDayjs(bar.value, "start"))) {
+      return
+    }
+    bar.value[barEnd.value] = newBarEnd
+    onDrag(e, bar.value)
   }
 
   const isOutOfRange = (xStart?: number, xEnd?: number) => {
-    if (!pushOnOverlap) {
+    if (!pushOnOverlap.value) {
```
Comment (InfectoOne) [inline src/composables/useBarDrag.ts:103]: I was convinced that I have fixed this already. Maybe I haven't pushed it to the master branch. In any case, thanks!

**Review by InfectoOne — CHANGES_REQUESTED** [review]
Excellent contribution. I have learned a few things reviewing this!
If you could just submit the changes noted in the comments, I'd accept the merge request.  
I have played around with the gantt chart and tested the changes out. No issues so far.

TLDR: `createBarDrag`should be reverted to `useBarDrag`. The new helper function `getBarElement` should be an arrow function.

File: src/composables/useBarDrag.ts (hunk 1 above)
Comment (dpschen) [reply src/composables/useBarDrag.ts:9]: The reason I renamed the function was that it isn't meant to be used in places where all the ['Usage Restriction'](https://vuejs.org/guide/reusability/composables.html#conventions-and-best-practices) for composables can be fulfilled. More specific: the useBarDrag / createBarDrag is currently only used and built for the usage inside of initDragOfBar. The latter function is only used inside of `firstMousemoveCallback`

I found this out the hard way because I wanted to inject the config in here as well.
This was not possible because access to an active component instance is necessary.

This led me to the conclusion that we do not have a 'real' composable here.

While writing this I realise that this might also lead to memory leaks like written in that part of the vue docs.

So whatever we conclude from this:
The name of the function and the file should be aligned :D

File: src/composables/useBarDrag.ts (hunk 3 above)
Comment (dpschen) [reply src/composables/useBarDrag.ts:103]: I only found it with TypeScript if I remember correctly :)

File: src/composables/useBarDrag.ts (hunk 2 above)
Comment (dpschen) [reply src/composables/useBarDrag.ts]: Fixed in 26c13c64d4a2a13e07ea1780924834ce3b77bc90

**Comment by InfectoOne** [comment]
Done! Thank you for your contribution, @dpschen !

**Comment by InfectoOne** [comment]
I will create a new build based on this version and publish it to npm.
Btw. I have noticed that you have bumped the version to 3.0.0. It'd be better to bump it from 2.0.4 to 2.1.0, since it's not a major release, but a minor release with a new minor functionality (the types are now exposed) and a lot of patches/fixes. Thoughts on this, @dpschen ?

**Comment by dpschen** [comment]
@InfectoOne 

Nice to hear!
I was not sure if I changed the API at some point in a wake that brakes stuff, so I better wanted to make sure to indicate bigger changes.
The 'css-api' did change though. So some manual style overwrites might break.

I didn't fix the [naming alignment](https://github.com/InfectoOne/vue-ganttastic/pull/53#discussion_r970799866) now, because I was waiting for your answer.

**Comment by InfectoOne** [comment]
No problem, I will fix the naming alignment myself. Technically speaking, the changes made to the CSS styling are 'breaking changes' to some extent, but manual CSS overwrites are often a "dirty/hacky" matter which if (slightly) broken does not justify a major release (in my opinion), so the version will be bumped to 2.1.0.
